File: project-{topic}-enhanced/src/middleware/logging.middleware.ts
```typescript
import { Request, Response, NextFunction } from 'express';
import { logger } from '../utils/logger';

/**
 * @file Request logging middleware.
 *
 * This middleware logs details about incoming HTTP requests and their responses,
 * including method, URL, status code, and response time.
 */

/**
 * Middleware to log details of each incoming request and its response.
 *
 * @param {Request} req - The Express request object.
 * @param {Response} res - The Express response object.
 * @param {NextFunction} next - The next middleware function.
 */
export const requestLogger = (req: Request, res: Response, next: NextFunction): void => {
  const startHrTime = process.hrtime.bigint(); // High-resolution time for timing requests

  res.on('finish', () => {
    const endHrTime = process.hrtime.bigint();
    const durationMs = Number(endHrTime - startHrTime) / 1_000_000; // Convert nanoseconds to milliseconds

    const logEntry = {
      method: req.method,
      url: req.originalUrl,
      statusCode: res.statusCode,
      response_time_ms: durationMs.toFixed(3),
      ip: req.ip,
      userAgent: req.get('User-Agent'),
      userId: (req as any).user ? (req as any).user.id : 'anonymous' // If auth middleware runs before this
    };

    if (res.statusCode >= 500) {
      logger.error('HTTP Request Error', logEntry);
    } else if (res.statusCode >= 400) {
      logger.warn('HTTP Request Warning', logEntry);
    } else {
      logger.info('HTTP Request', logEntry);
    }
  });

  next();
};
```

File: project-{topic}-optimized/app/middleware/logging_middleware.py
import logging
import time

from starlette.middleware.base import BaseHTTPMiddleware
from starlette.responses import Response
from starlette.types import ASGIApp

logger = logging.getLogger(__name__)

class LoggingMiddleware(BaseHTTPMiddleware):
    """
    Custom middleware for request logging.
    Logs incoming requests and outgoing responses, including processing time.
    """
    def __init__(self, app: ASGIApp):
        super().__init__(app)

    async def dispatch(self, request, call_next):
        start_time = time.time()
        
        # Log request details
        logger.info(f"Request: {request.method} {request.url.path} from {request.client.host}")

        response = await call_next(request)

        process_time = time.time() - start_time
        # Log response details and processing time
        logger.info(f"Response: {request.method} {request.url.path} Status: {response.status_code} "
                    f"Process Time: {process_time:.4f}s")
        
        return response
```